### PR TITLE
CFSQL-1294 be more faithful to sqlite's algorithm for printing newlines in export

### LINKED ISSUE #9042: D1 export by Wrangler did not escape the literal `\n` properly
### Which Cloudflare product(s) does this pertain to?

D1

### What versions & operating system are you using?

Wrangler 4.13.0

### Please provide a link to a minimal reproduction

_No response_

### Describe the Bug

* Create a table with a text column, and add some data with both the literal `\n` and normal linebreaks
* Export the database with the Wrangler command `wrangler d1 export <db_name> --remote --output=./test.sql`
* Inspect the SQL file downloaded

Image

```sql
PRAGMA defer_foreign_keys=TRUE;
CREATE TABLE [test] ("test_text" text);
INSERT INTO test VALUES('AAA\nAAA');
INSERT INTO test VALUES(replace('AAA\nAAA','\n',char(10)));
INSERT INTO test VALUES(replace('AAA\nAAA\nBBB','\n',char(10)));
```

In this table, the first record contains only the literal `\n`, the second one contains only linebreaks, these two appear intact.
However, the third one was messed up, and the literal `\n` was converted to linebreaks.

### Please provide any relevant error logs

_No response_

### PR DESCRIPTION
**Pull request by alsuren**
If newline and backslash n are both taken, it picks a new string to use

Fixes cloudflare/workers-sdk#9042

_Describe your change..._

Ports the sqlite3 .dump algorithm a bit more faithfully, by checking for collisions in proposed escape sequences and then generating new escape sequences until it finds one that doesn't collide.

Extended tests pass in internal d1-worker copy of this change.

---

- Tests
  - [ ] Tests included/updated
  - [ ] Automated tests not possible - manual testing has been completed as follows:
  - [x] Additional testing not necessary because: tests for this file live in the internal d1-worker repo
- Public documentation
  - [ ] Cloudflare docs PR(s): 
  - [x] Documentation not necessary because: bugfix of edge case in the export flow.

*A picture of a cute animal (not mandatory, but encouraged)*

Screenshot 2026-08-24 at 13 53 42

https://freesvg.org/rabbit-line-art (I hope you like what I did there :D)

---

<a href="https://cloudflare.devinenterprise.com/review/cloudflare/workers-sdk/pull/15334" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=3">
    Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>

- .changeset/two-lamps-leave.md: [@cloudflare/wrangler]
- packages/miniflare/src/workers/d1/dumpSql.ts: [@cloudflare/wrangler]

</details>

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>

- .changeset/two-lamps-leave.md: [@cloudflare/wrangler]
- packages/miniflare/src/workers/d1/dumpSql.ts: [@cloudflare/wrangler]

</details>

File: packages/miniflare/src/workers/d1/dumpSql.ts [hunk 1]
```diff
@@ -40,7 +40,7 @@ export function* dumpSql(
 	const { noData, noSchema } = options || {};
 
 	// Taken from SQLite shell.c.in https://github.com/sqlite/sqlite/blob/105c20648e1b05839fd0638686b95f2e3998abcb/src/shell.c.in#L8463-L8469
-	// @ts-expect-error -- SqlStorageStatement needs to be callable
+	// @ts-ignore (SqlStorageStatement needs to be callable)
```
Comment (dario-piotrowicz) [inline packages/miniflare/src/workers/d1/dumpSql.ts:43]: why this change? generally `@ts-expect-error` is preferred

**Review by dario-piotrowicz — COMMENTED** [review]
Generally looks ok to me, besides the one small thing I commented on

Also the CI seems quite red and there's a merge conflict 🥲

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>

- .changeset/two-lamps-leave.md: [@cloudflare/wrangler]
- packages/miniflare/src/workers/d1/dumpSql.ts: [@cloudflare/wrangler]

</details>

File: packages/miniflare/src/workers/d1/dumpSql.ts (hunk 1 above)
Comment (alsuren) [reply packages/miniflare/src/workers/d1/dumpSql.ts]: Oh I see what might happened. This file is vendored into the d1 codebase, and I fixed it there first. I then just copied the file over here and reverted all of the changes that were made by others.

It seems like there are other changes, so I will need to copy them back over at some point.

At the time this file was written, there wasn't a good package to put a file that could be imported into the d1-worker durable object code, so the file was vendored in both places and kept in sync by hand (or not, as I found out when I joined and changes had been made in both places).

I feel like things might have changed since then. Is there somewhere that we could move this file so that we can import it from our DO code and also import it from miniflare, to avoid this?

File: packages/miniflare/src/workers/d1/dumpSql.ts [hunk 2]
```diff
@@ -201,38 +201,51 @@ export function* dumpSql(
...
+		.replaceAll("'", "''")
+		.replaceAll("\n", newlinePlaceholder)
+		.replaceAll("\r", carriageReturnPlaceholder);
+	let output = `'${escaped}'`;
+	if (hasCarriageReturn) {
+		output = `replace(${output},'${carriageReturnPlaceholder}',char(13))`;
+	}
+	if (hasNewline) {
+		output = `replace(${output},'${newlinePlaceholder}',char(10))`;
+	}
```
[the remark this answers is not in the document]
Comment (alsuren) [reply packages/miniflare/src/workers/d1/dumpSql.ts:231]: The copy in the d1-worker repo had tests added when it was implemented as part of CFSQL-1294. I'd rather not maintain two test suites for the same piece of code.

**Review by FlorentCollin — APPROVED** [review]
